src/api/cpp/array.cpp

```diff
@@ -410,54 +443,59 @@ namespace af
-        af_seq afs[4];                                                          \
-        getSeq(afs);                                                            \
-        af::dim4 cdims = isRef ? seqToDims(afs, this->dims()) : this->dims();   \
-        array cst = constant(value, cdims, this->type());                       \
-        return operator op(cst);                                                \
-    }                                                                           \
-    array& array::operator op(const cdouble &value)                             \
-    {                                                                           \
-        af_seq afs[4];                                                          \
-        getSeq(afs);                                                            \
-        af::dim4 cdims = isRef ? seqToDims(afs, this->dims()) : this->dims();   \
-        array cst = constant(value, cdims);                                     \
-        return operator op(cst);                                                \
-    }                                                                           \
-    array& array::operator op(const cfloat &value)                              \
-    {                                                                           \
-        af_seq afs[4];                                                          \
-        getSeq(afs);                                                            \
-        af::dim4 cdims = isRef ? seqToDims(afs, this->dims()) : this->dims();   \
-        array cst = constant(value, cdims);                                     \
-        return operator op(cst);                                                \
-    }                                                                           \
-
-    INSTANTIATE(+=, +, af_add)
-    INSTANTIATE(-=, -, af_sub)
-    INSTANTIATE(*=, *, af_mul)
-    INSTANTIATE(/=, /, af_div)
+#define INSTANTIATE(op, op1)                                            \
+    array& array::operator op(const array &other)                       \
+    {                                                                   \
+        bool this_ref = isRef;                                          \
+        if (this_ref) {                                                 \
+            af_array lhs;                                               \
+            int dim = gforDim(this->s);                                 \
+            AF_THROW(af_weak_copy(&lhs, this->arr));                    \
+            unsigned ndims = numDims(lhs);                              \
+            /* FIXME: Unify with other af_assign */                     \
+            array tmp = *this op1 other;                                \
+            af_seq afs[4];                                              \
+            getSeq(afs);                                                \
```

Comment:
hm, `this->s` isn't changing after line 451 and `getSeq(afs)` will get us same `this->s`, isn't it ?